### PR TITLE
Neutrino Id Training Purity & Completeness

### PR DESCRIPTION
**Pull request by StevenGreen1**
Modifications to purity and completeness calculation in neutrino id training.

There is potential for a merge conflict in the LArMCParticleHelper.h file with the Bdt pull request (https://github.com/PandoraPFA/LArContent/pull/42) where the same change was made (making CollectReconstructable2DHits a public function).  This is the only change though and should be easy to fix up with some rebasing.  

This was tested on a few neutrino + cosmic events and the changes were minimal.  The purities and completenesses did change slightly, but in all cases seen the best slice index was unchanged. 

@a-d-smith If you have time to have a quick look over this to make sure there are no obvious issues that'd be great.  Thanks.

### CODE REVIEW

**Comment by a-d-smith** [comment]
Looks good to me, the only thing I would say is the LArMCParticleHelper::PrimaryParameters already has default values hardcoded:
https://github.com/PandoraPFA/LArContent/blob/133054eb310b215a18a1cae2f74d68c9f12d1d6e/larpandoracontent/LArHelpers/LArMCParticleHelper.cc#L31-L39

Maybe instead of doing this twice, the constructor could use these values instead?
What do you think?

**Comment by StevenGreen1** [comment]
Thanks for looking over this.  Yeah that's a good idea, I'll make that change now.

File: larpandoracontent/LArControlFlow/NeutrinoIdTool.cc [hunk 1]
```diff
@@ -141,6 +158,9 @@ void NeutrinoIdTool::Collect2DHits(const PfoList &pfos, CaloHitList &hitList) co
         // ATTN hits collected from Pfos are copies of hits passed from master instance, we need to access their parent to use MC info
         const CaloHit *pParentCaloHit(static_cast<const CaloHit *>(pCaloHit->GetParentAddress()));
 
+        if (std::find(reconstructableCaloHitList.begin(), reconstructableCaloHitList.end(), pParentCaloHit) == reconstructableCaloHitList.end())
+            continue;
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/NeutrinoIdTool.cc]: Consider use of CaloHitSet (note: unordered!) here for lookup.

File: larpandoracontent/LArControlFlow/NeutrinoIdTool.cc [hunk 2]
```diff
@@ -167,6 +187,8 @@ unsigned int NeutrinoIdTool::CountNeutrinoInducedHits(const CaloHitList &hitList
     return nNuHits;
 }
 
+
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/NeutrinoIdTool.cc]: Two empty lines! Two!

Comment (StevenGreen1) [reply larpandoracontent/LArControlFlow/NeutrinoIdTool.cc]: I thought it helped the fung shui of the code, but happy to remove.

File: larpandoracontent/LArControlFlow/NeutrinoIdTool.cc [hunk 3]
```diff
@@ -158,21 +159,21 @@ void NeutrinoIdTool::Collect2DHits(const PfoList &pfos, CaloHitList &hitList, Ca
...
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-unsigned int NeutrinoIdTool::CountNeutrinoInducedHits(const CaloHitList &hitList) const
+unsigned int NeutrinoIdTool::CountNeutrinoInducedHits(const CaloHitSet &caloHitSet) const
 {
     unsigned int nNuHits(0);
-    for (const CaloHit *const pCaloHit : hitList)
+    for (const CaloHit *const pCaloHit : caloHitSet)
     {
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/NeutrinoIdTool.cc]: As discussed, use unordered_set for the look-ups, but list (or vector) for ordered loops (making sure that contents of list of vector hadn't been in a set as n intermediate step!).

**Comment by StevenGreen1** [comment]
I've tested these changes with a few events and the reported purity and completenesses appear as you would expect.

**Pull request merged**